README.md

```diff
@@ -25,464 +24,423 @@ Features include:
+	responses.BadRequest(),
+).Run(func(ctx huma.Context) {
+	for i := 0; i < 5; i++ {
+		// Use AddError to add multiple error details to the response.
+		ctx.AddError(fmt.Errorf("Error %d", i))
+	}
+
+	// Check if the context has had any errors added yet.
+	if ctx.HasError() {
+		// Use WriteError to set the actual status code, top-level message, and
+		// any additional errors. This sends the response.
+		ctx.WriteError(http.StatusBadRequest, "Bad input")
+		return
+	}
 })
 ```
 
-You can also declare parameters with additional validation logic by using the `schema` module:
+## Request Inputs
+
+Requests can have parameters and/or a body as input to the handler function. Like responses, inputs use standard Go structs but the tags are different. Here are the available tags:
+
+| Tag      | Description                        | Example                  |
+| -------- | ---------------------------------- | ------------------------ |
+| `path`   | Name of the path parameter         | `path:"thing-id"`        |
+| `query`  | Name of the query string parameter | `query:"q"`              |
+| `header` | Name of the header parameter       | `header:"Authorization"` |
+
+The following types are supported out of the box:
+
+| Type                | Example Inputs         |
+| ------------------- | ---------------------- |
+| `bool`              | `true`, `false`        |
+| `[u]int[16/32/64]`  | `1234`, `5`, `-1`      |
+| `float32/64`        | `1.234`, `1.0`         |
+| `string`            | `hello`, `t`           |
+| `time.Time`         | `2020-01-01T12:00:00Z` |
+| slice, e.g. `[]int` | `1,2,3`, `tag1,tag2`   |
+
+For example, if the parameter is a query param and the type is `[]string` it might look like `?tags=tag1,tag2` in the URI.
+
+The special struct field `Body` will be treated as the input request body and can refer to another struct or you can embed a struct inline.
+
+Here is an example:
 
 ```go
-s := schema.Schema{
-	MinLength: 1,
-	MaxLength: 32,
+type MyInputBody struct {
+	Name string `json:"name"`
+}
+
+type MyInput struct {
+	ThingID     string      `path:"thing-id" doc:"Example path parameter"`
+	QueryParam  int         `query:"q" doc:"Example query string parameter"`
+	HeaderParam string      `header:"Foo" doc:"Example header parameter"`
+	Body        MyInputBody `doc:"Example request body"`
 }
 
-huma.PathParam("id", "Note ID", huma.Schema(s))
+// ... Later you use the inputs
+
+// Declare a resource with a path parameter that matches the input struct. This
+// is needed because path parameter positions matter in the URL.
+thing := app.Resource("/things/{thing-id}")
+
+// Next, declare the handler with an input argument.
+thing.Get("get-thing", "Get a single thing",
+	responses.NoContent(),
+).Run(func(ctx huma.Context, input MyInput) {
+	fmt.Printf("Thing ID: %s\n", input.ThingID)
+	fmt.Printf("Query param: %s\n", input.QueryParam)
+	fmt.Printf("Header param: %s\n", input.HeaderParam)
+	fmt.Printf("Body name: %s\n", input.Body.Name)
```

Comment:
I agree this is somewhat different from those other frameworks, and for the most part this is on purpose despite it introducing a (hopefully small) learning curve. Parameters are a good example, so let's take the Express example because I love that framework:

```js
app.get('/user/:user', function(req, res, next){
  res.send('user ' + req.user.name);
});
```

What type is `:user`? What sort of validation is on that parameter? Is there any documentation for it? Is it an ID, a name, something else? If you look earlier in the code you see this:

```js
app.param('user', function(req, res, next, id){
  if (req.user = users[id]) {
    next();
  } else {
    next(createError(404, 'failed to find user'));
  }
});
```

So there is some validation actually, and a very non-obvious possibility of getting a 404 outside of the handler itself. Outside of *maybe* really complex static code analysis you have no way of determining this is the case. You couldn't generate documentation off it. It's not self-descriptive or obvious.

Part of Huma's reason for existing is that I'm saying the above doesn't meet my minimum bar for maintainable, self-descriptive service code or a user-friendly client experience unless they also have hand-written up-to-date docs somewhere. In Huma we try to make it much more obvious, and it does use the param in the request URI template because we need to know the position of it:

```go
type UserParam struct {
  User string `path:"user-name" pattern="^[a-z0-9-]+$" doc:"Unique user name"`
}

app := cli.NewRouter("Test Router", "1.0.0")

app.Resource("/users/{user-name}").Get("get-user", "Get a user by name",
  responses.OK().Model(User),
  responses.NotFound(),
).Run(func(ctx huma.Context, input UserParam) {
  u, err := db.GetUser(input.User)
  if err != nil {
    ctx.WriteError(http.StatusNotFound, input.User + " user not found")
    return
  }

  ctx.WriteModel(http.StatusOK, u)
})
```

Now we get response info, param type information & safety, validation through a regex, documentation for OpenAPI (docs, mocks, SDKs, CLIs, etc), and hopefully straightforward code out of the box. My hope is that these things are worth the learning curve coming from other frameworks.